src/taskgraph/run-task/run-task

```diff
@@ -705,23 +706,40 @@ def git_checkout(
 
     run_required_command(b"vcs", args, cwd=destination_path)
 
-    if os.path.exists(os.path.join(destination_path, ".gitmodules")):
-        args = [
-            "git",
-            "submodule",
-            "init",
-        ]
-
-        run_required_command(b"vcs", args, cwd=destination_path)
-
-        args = [
-            "git",
-            "submodule",
-            "update",
-            "--force",  # Overrides any potential local changes
+    if submodules is not None:
+        def _filter_submodule(status):
+            # A boolean value selects/disables all submodules.
+            if isinstance(submodules, bool):
+                return submodules
+
+            # Otherwise, a colon-separated stringlist of selected submodules.
+            smpath = status.split()[1]
+            return smpath in submodules.split(':')
+
+        submodpaths = [
+            status.split()[1]
+            for status in subprocess.check_output(["git", "submodule", "status"],
+                cwd=destination_path, universal_newlines=True).splitlines()
+            if _filter_submodule(status)
         ]
 
-        run_required_command(b"vcs", args, cwd=destination_path)
+        for p in submodpaths:
+            args = [
+                "git",
+                "submodule",
+                "init",
+                p
```

Comment:
Checkouts are often cached on individual workers. Will this work correctly if a checkout already contains an initialized submodule at the same (or different) revision?

Reply:
I think it would be technically possible to have different submodules checked out at different revisions, but technically a task should only rely on the submodules that it declares, and those would all get the `git submodule update` treatment below.

That being said, I'm starting to think we should default to updating all submodules if `SUBMODULES` is not set. I don't think it's going to be easy to audit tasks to tell whether / which submodules they may or may not be using. So preserving backwards compatibility by defaulting to all of them will probably save us some headache.

Reply:
It will take a little bit of rethinking how what the env variables should mean to properly convey that (default disabled is easy because an unset environment variable is read as an empty string), but I will see what I can do.